kratos/utilities/math_utils.h

```diff
@@ -502,15 +502,63 @@ class MathUtils
         {
             Det = Det2(A);
         }
-        else
+        else if (A.size1() == 3)
         {
             Det = Det3(A);
         }
+        else if (A.size1() == 4)
+        {
+            Det = Det4(A);
+        }
+        else
+        {
+            using namespace boost::numeric::ublas;
+            typedef permutation_matrix<std::size_t> pmatrix;
+            Matrix Aux(A);
+            pmatrix pm(Aux.size1());
+            bool singular = lu_factorize(Aux,pm);
+            
+            if (singular == true)
+            {
+                return 0.0;
+            }
+            
+            Det = 1.0;
+            
+            for (unsigned int i = 0; i < Aux.size1();i++)
+            {
+                unsigned int ki = pm[i] == i ? 0 : 1;
+                Det *= std::pow(-1.0, ki) * Aux(i,i);
+            }
+        }
 
         return Det;
     }
     
     /**
+     * Calculates the determinant of a matrix of dimension 2x2 or 3x3 (no check performed on matrix size)
+     * @param A: Is the input matrix
+     * @return The determinant of the 2x2 matrix
+     */
+        
+    static inline TDataType GeneralizedDet(const MatrixType& A)
+    {
+        TDataType Determinant;
+        
+        if (A.size1() == A.size2())
+        {
+            Determinant = Det(A);
```

Comment:
I don't understand your suggestion, the thing is to keep the current determinant calculation, and in case that the Jacobian is not square let the user use the generalized determinant.